### PR TITLE
Feature/compose

### PR DESCRIPTION
**Pull request by mpetuska**
Upgrades versions and adds new redux-kotlin-compose module

Usage:
```


data class MyState(val name: String, val sureName: String)
sealed class MyAction {
  object Anonymize : MyAction()
}

val myReducer: Reducer<MyState> = { state, action ->
  when (action) {
    is MyAction.Anonymize -> state.copy(sureName = "N/A")
    else -> state
  }
}
val myStore = createStore(myReducer, MyState("Mike", "Kox"))

inline fun <TSlice> selectMyState(
  crossinline selector: @DisallowComposableCalls MyState.() -> TSlice
): State<TSlice> = selectState(selector)

@Composable
fun MyApp() {
  withStore(myStore) {
    val dispatcher = rememberDispatcher()
    val surename by selectMyState { sureName }
    Text(surename)
    Button(attrs = { onClick {
      dispatcher(MyAction.Anonymize)
    } })
  }
}
```

### CODE REVIEW

**Comment by mpetuska** [comment]
Some more polishing on the way (including formatting), but I wanted to raise a PR early to get some feedback on API design.

**Comment by mpetuska** [comment]
PR is now ready for review.

**Comment by Syrou** [comment]
@patjackson52 Any chance of getting this merged? It would be a great feature in my book

**Comment by mpetuska** [comment]
Lol I totally forgot I had this PR raised 😆

**Comment by patjackson52** [comment]
> @patjackson52 Any chance of getting this merged? It would be a great feature in my book

Yes, thanks for the ping.  It looks like it needs some tweaks.  KtLint is failing. @mpetuska can you take a look?

Any details on this book?

**Comment by mpetuska** [comment]
It's complaining about something that I'm sure is correct
```
/github/workspace/redux-kotlin-compose/src/commonMain/kotlin/LocalStore.kt:27:85: Not a valid Kotlin file (type expected) (cannot be auto-corrected)
```

Refers to this piece of code
```
@Composable
inline fun <T : Any> withStore(store: Store<T>, crossinline content: @Composable () -> Unit) {
  CompositionLocalProvider(LocalStore provides store) { content() }
}
```

Looks like a bug in ktlint when faced with generics and inline functions.

Also, what book?

**Comment by Syrou** [comment]
@mpetuska Does this not introduce a problem for iOS targets if added to common?

**Comment by mpetuska** [comment]
i.e. it's only available for jvm, js and android

**Comment by Syrou** [comment]
Yeah figured, so the redux-kotlin-compose dependency can't be added to commonMain like the rest of the things in this project, it is understandable, but quite a bummer.

**Comment by mpetuska** [comment]
It can if you only target jvm js and android

**Comment by mpetuska** [comment]
@patjackson52 could we get this merged in February? Happy to do as many revisions as needed.

Or would you prefer to wait for compose-jb 1.1.0 release instead?

**Comment by patjackson52** [comment]
> 

I'd love to get this in this month.  Pretty swamped with other things so not sure when.  If someone else is interested in driving this and a bit more responsibility with the repo we can talk

**Comment by junerver** [comment]
This is a great PR, with it I think I can use Compose in my android app with great confidence

**Comment by Syrou** [comment]
@mpetuska @junerver Perhaps it is time to just fork this project and start adding the PR's and fixes. https://github.com/1gravity/redux-kotlin is the most recent fork that fixed up a lot of issues perhaps we should start our efforts there?

**Comment by mpetuska** [comment]
Is the fork published to maven central?

**Comment by Syrou** [comment]
> Is the fork published to maven central?

Seems like it yeah

**Comment by mpetuska** [comment]
Quite frankly, I believe it would be best for someone to step-up as maintainer of this repo instead of forking it. Much smoother experience for the end-users

**Comment by Syrou** [comment]
> Quite frankly, I believe it would be best for someone to step-up as maintainer of this repo instead of forking it. Much smoother experience for the end-users

Sure, it would be the best. It all depends if @patjackson52 picks this up or not, to assign someone else. It is not the first time forks has taken over though.

**Comment by patjackson52** [comment]
> > Quite frankly, I believe it would be best for someone to step-up as maintainer of this repo instead of forking it. Much smoother experience for the end-users
> 
> Sure, it would be the best. It all depends if @patjackson52 picks this up or not, to assign someone else. It is not the first time forks has taken over though.

Ideally we could keep the same repo and keep moving with this.  Wish I had more time to commit to this, but just not possible.  Perhaps we can set up a chat, discord, group somewhere to discuss.  Is the main concern here publishing artifacts?

**Comment by mpetuska** [comment]
@patjackson52 I thought this through, and if you're happy with it, I could step up as a maintainer in a limited capacity (reviewing PRs, keeping versions up to date, critical patches) at least temporarily.

**Comment by Syrou** [comment]
+1 for making @mpetuska maintainer. Seing the activity of the other projects he has It feels like you'd put this project into safe hands

**Comment by patjackson52** [comment]
> @patjackson52 I thought this through, and if you're happy with it, I could step up as a maintainer in a limited capacity (reviewing PRs, keeping versions up to date, critical patches) at least temporarily.

@mpetuska yes, that could help, thank you for the offer.

**Comment by mpetuska** [comment]
@patjackson52 Could you merge this PR with admin permissions, ktlint can't handle compose unfortunatelly and reports false-positives. I'll raise another PR afterwards to switch it out to detekt managed ktlint to avoid such false-positives in the future.

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,4 +1,9 @@
 ## Unreleased
+- Kotlin 1.6.10
+- Gradle 7.3.2
+- Coroutines 1.5.2
+- Added refreshVersions plugin
+- New redux-kotlin-compose module
```
Comment (patjackson52) [inline CHANGELOG.md:6]: lots of changes for 1 commit.  
Ideally these are broken into their own commits.

File: build.gradle.kts [hunk 2]
```diff
@@ -1,38 +1,10 @@
...
-        maven("https://oss.sonatype.org/content/repositories/snapshots")
-        mavenCentral()
-    }
-
-    group = project.properties["GROUP"]!!
-    version = project.properties["VERSION_NAME"]!!
-    if (hasProperty("SNAPSHOT") || System.getenv("SNAPSHOT") != null) {
-        version = "$version-SNAPSHOT"
-    }
+gradleEnterprise {
```
Comment (patjackson52) [inline build.gradle.kts:5]: I've been away from gradle for awhile - what's the deal with `gradleEnterprise`?

**Comment by patjackson52** [comment]
Apologies for the inactivity and delayed review - This is an awesome addition, thanks for the PR.
Only major concern I have is I'm not convinced this belongs in the 'core' repo & artifact.  Seems like more of an addition specific to Compose.  Not all users of redux-kotlin use compose.   Seems more appropriate to put in a separate repo & artifact (we can add repo in the reduxkotlin 'org' on github).   This was deliberately modularized to keep the core of reduxkotlin agnostic of any platform/framework.  If I'm missing something, convince me!

Otherwise all the formatting improvements are awesome and should stay. @mpetuska

**Comment by junerver** [comment]
> Apologies for the inactivity and delayed review - This is an awesome addition, thanks for the PR. Only major concern I have is I'm not convinced this belongs in the 'core' repo & artifact. Seems like more of an addition specific to Compose. Not all users of redux-kotlin use compose. Seems more appropriate to put in a separate repo & artifact (we can add repo in the reduxkotlin 'org' on github). This was deliberately modularized to keep the core of reduxkotlin agnostic of any platform/framework. If I'm missing something, convince me!
> 
> Otherwise all the formatting improvements are awesome and should stay. @mpetuska

I agree with this point of view, redux-kotlin is like redux, feature/compose is like react-redux, it is used to connect redux-kotlin with Android Jetpack Compose, it should be a separate repo. feature/compose is necessary, but it should not be in the core library

**Comment by mpetuska** [comment]
Fair enough. I personally prefer integration modules to be within the same repo to make keeping dependencies up to date easier and also have it share the same version with core lib. But that's mainly preference and not necessarily a "best" way to do things. I'll be closing this PR in favour of the new repo.

**Comment by mpetuska** [comment]
The stage is set https://github.com/reduxkotlin/redux-kotlin-compose

File: build.gradle.kts (hunk 2 above)
Comment (mpetuska) [reply build.gradle.kts:5]: This basically allows build scans in CI which makes it a lot easier to debug issues

**Pull request closed without merging**